### PR TITLE
Implement a cheap payload "does that look like json" check

### LINKED ISSUE #61: Cheap checks on uploaded data
This is a quick stopgap for #20.

In this issue we want to do small checks that can be done while receiving the data.
The small check I have in mind is to make sure that the first character is a `{`.

### PR DESCRIPTION
**Pull request by julienw**
Fixes #61

The patch isn't so big but the stream API is confusing so I expect that not everything is straightforward. I tried to add comments to store my hard earned knowledge in the source files, for my future myself and coworkers, but please ask if you see anything that's unclear.

Also I split it in separate commits, with tests, so hopefully this should be easier to read by looking at the individual commits. (last commit is https://github.com/firefox-devtools/profiler-server/pull/85, please ignore it here)

I also tested uploading from the profiler UI, this was working properly.

### CODE REVIEW

File: src/routes/publish.js [hunk 1]
```diff
@@ -42,29 +45,78 @@ export function publishRoutes() {
...
       lengthChecker,
       hasherTransform,
       concatener
     );
 
+    // 2. "Gunzip and check" json pipeline
+    // This pipeline will stop quite early, once we know this looks like a json.
+    // When it ends we'll just unpipe.
+    // We don't use the `pipeline` utility because with it we get some
+    // "premature close" errors when the length checker errors out.
```
Comment (julienw) [inline src/routes/publish.js:85]: Frankly I don't understand why we get these errors here but not in the previous pipeline... I spent too much time understanding all kind of errors and I'm a bit fed up with this, so I didn't want to invest more time now that everything works fine with the current implementation... I'm a bit frustrated that I don't understand this, so maybe I'll come back to it later, but I want to move on :-D

File: src/types/libdef/npm/jest_v25.x.x.js [hunk 2]
```diff
@@ -901,7 +901,7 @@ type JestObjectType = {
 type JestSpyType = { calls: JestCallsType, ... };
 
 type JestDoneFn = {|
- (): void,
+ (error?: Error): void,
```
Comment (julienw) [inline src/types/libdef/npm/jest_v25.x.x.js:904]: The change has been contributed to flow-typed: https://github.com/flow-typed/flow-typed/pull/3844 so there will be no problem updating the libdef later.

File: src/utils/streams.js [hunk 3]
```diff
@@ -25,7 +28,7 @@ export class HasherPassThrough extends Transform {
 
   _transform(
     chunk: string | Buffer,
-    encoding: mixed,
+    encoding: string,
```
Comment (julienw) [inline src/utils/streams.js:31]: This was a mistake from previous patches.

File: src/utils/streams.js [hunk 4]
```diff
@@ -36,7 +39,7 @@ export class HasherPassThrough extends Transform {
   // This method returns the computed sha1 from the data that passed through the
   // transform.
   sha1(): string {
-    this.log.verbose('sha1()');
+    this.log.trace('sha1()');
```
Comment (julienw) [inline src/utils/streams.js:42]: Using `trace` generates a stack automatically which is nice when debugging. To see trace logs we need to run tests with `LOG_LEVEL=trace`, they don't show up when the log level is "verbose", so it's also more silent.

File: src/utils/streams.js [hunk 5]
```diff
@@ -110,18 +113,26 @@ export class Concatenator extends Writable {
...
+    this.log.trace('_destroy()');
     this.chunks.length = 0;
     this.contents = null;
-    callback();
+
+    // Passthrough the error information, if present.
+    // This line is needed because of the slightly inconsistent
+    // signature of callback vs err, and that we can't change.
+    err = err || undefined;
+    callback(err);
```
Comment (julienw) [inline src/utils/streams.js:124]: This was also a mistake from my side: we're supposed to forward the error to the callback. This API is so confusing and the documentation is missing some information...

File: src/utils/streams.js [hunk 6]
```diff
@@ -130,3 +141,173 @@ export class Concatenator extends Writable {
...
+    }
+
+    // If we're coming here, this means we never finished checking. Let's
+    // happily throw, then!
+    callback(new BadRequestError(this.errorMessage));
+  }
+}
+
+// This simple wrapper simply encapsulates the native gunzip stream and rewrite
+// errors to make them more compatible with our code and koa's error handling.
```
Comment (julienw) [inline src/utils/streams.js:236]: So much code just to have nice errors :( I'm a bit sad.

File: src/utils/streams.js [hunk 7]
```diff
@@ -279,3 +279,35 @@ export class GunzipWrapper extends Transform {
     callback(err);
   }
 }
+
+// This tool simply forward an error happening on a stream to other streams, by
+// destroying them.
+export function forwardErrors(
```
Comment (julienw) [inline src/utils/streams.js:285]: I didn't add unit tests for this utility but I believe the tests for the publishing cover it, by ensuring we get the right errors when feeding some bad payload to the server.

File: .circleci/config.yml [hunk 8]
```diff
@@ -3,7 +3,7 @@ version: 2.1
 executors:
   node:
     docker:
-      - image: circleci/node:10.15
+      - image: cimg/node:10.21
```
Comment (julienw) [inline .circleci/config.yml:6]: This change has been extracted to https://github.com/firefox-devtools/profiler-server/pull/85, it's in this PR just so that the tests pass.

File: bin/pre-install.js [hunk 9]
```diff
@@ -104,9 +104,7 @@ function parseExpectedNodeVersion() {
   const circleConfig = fs.readFileSync('.circleci/config.yml', {
     encoding: 'utf8',
   });
-  const expectedNodeVersion = /image: circleci\/node:([\d.]+)/.exec(
-    circleConfig
-  );
+  const expectedNodeVersion = /image: cimg\/node:([\d.]+)/.exec(circleConfig);
```
Comment (julienw) [inline bin/pre-install.js:107]: This change has been extracted to #85, it's in this PR just so that the tests pass.

File: test/api/publish.test.js [hunk 10]
```diff
@@ -128,24 +130,18 @@ describe('publishing endpoints', () => {
...
   });
 
-  it('returns an error when the sent data is bigger than the length', async () => {
-    jest.spyOn(process.stdout, 'write').mockImplementation(() => {});
-    const req = getPreconfiguredRequest();
-    await req.set('Content-Length', String(3)).send('aaaa').expect(400); // 400 means Bad Request. It's generated automatically by Koa.
-    expect(process.stdout.write).toHaveBeenCalledWith(
-      expect.stringContaining('server_error')
-    );
-  });
```
Comment (julienw) [inline test/api/publish.test.js:138]: I remove this test case because:
* I couldn't get the error I wanted with the changes
* this is auto-handled by node and koa -- when I tested, I got results like 400, 500 or 405 (method not allowed). The underlying node error is actually "invalid method" (see https://github.com/nodejs/node/issues/33728 for more explanation about that)
* this is quite an edge case so I didn't feel it was important to invest time to make a test for it after all.

File: test/unit/streams.test.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  it('accepts normal content', async () => {
+    const fixture = '{ "foo": "bar" }';
+    const checker = new CheapJsonChecker();
+    const input = new PassThrough();
+    input.write(fixture.slice(0, 3));
+    input.end(fixture.slice(3));
+    await expect(pipeline(input, checker)).resolves.toBe(undefined);
+  });
+
+  it('accepts content with some space at the start', async () => {
```
Comment (gregtatum) [inline test/unit/streams.test.js:29]: Does this actually happen with our profiles?

File: src/routes/publish.js [hunk 12]
```diff
@@ -42,29 +45,78 @@ export function publishRoutes() {
...
     // error. However the HTTP request isn't part of the pipeline because we
     // don't want to destroy it. Indeed if it's destroyed, the socket is closed
     // and we can't send a nice error to the caller.
-    const pipeline = util.promisify(Stream.pipeline);
+    const pipeline = promisify(Stream.pipeline);
+    const finished = promisify(Stream.finished);
+
+    // We create 2 interconnected pipelines:
+    //
+    //     request
```
Comment (gregtatum) [inline src/routes/publish.js:62]: Thanks for the helpful diagram!

File: src/routes/publish.js (hunk 1 above)
Comment (gregtatum) [reply src/routes/publish.js:85]: Works for me, it's a bit unfortunate the streams API is so poorly designed.

File: src/routes/publish.js [hunk 13]
```diff
@@ -42,29 +45,78 @@ export function publishRoutes() {
...
+      // stop flowing once the buffers in the second pipeline are full.
+      lengthChecker.unpipe(gunzipStream);
+      gunzipStream.destroy();
+    });
+
+    // Forward errors between streams in the second pipeline, as well as the
+    // first pipeline through lengthChecker.
+    // Note that errors inside the first pipeline are forwarded thanks to the
+    // `pipeline` thingie.
+    forwardErrors(lengthChecker, gunzipStream, jsonChecker);
```
Comment (gregtatum) [inline src/routes/publish.js:108]: Wow, this feels like magic. (To be clear, because the streams API is magical)

File: test/unit/streams.test.js (hunk 11 above)
Comment (julienw) [reply test/unit/streams.test.js:29]: Not when feeding directly from the UI but I'm also thinking of some more "manual" use cases where profiles could be posted (for example from CI). Maybe too much precaution from my side...

File: src/routes/publish.js (hunk 13 above)
Comment (julienw) [reply src/routes/publish.js:108]: Forwardum Errorus
